ckanext/canada/harvesters.py

```diff
@@ -0,0 +1,616 @@
+    'num_tags', 'num_resources', 'maintainer',
+    'isopen', 'relationships_as_object', 'license_title',
+    'license_title_fra', 'license_url_fra', 'license_url',
+    'author',
+    'groups', # not used
+    'relationships_as_subject', 'department_number',
+    # FIXME: remove these when we can:
+    'resource_type',
+    'creator_user_id']
+
+SYNC_RESOURCE_TRIM_FIELDS = ['package_id', 'revision_id',
+    'revision_timestamp', 'cache_last_updated',
+    'webstore_last_updated', 'state',
+    'description', 'tracking_summary', 'mimetype_inner',
+    'mimetype', 'cache_url', 'created', 'webstore_url',
+    'position', 'metadata_modified']
+
+
+class PortalSync(plugins.SingletonPlugin):
+    """
+    Harvester class to sync Portal Datasets, Resources, Views,
+    and DataStores from the Registry.
+    """
+    plugins.implements(IHarvester)
+
+
+    _save_gather_error = HarvestGatherError.create
+    _save_object_error = HarvestObjectError.create
+
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        if not plugins.plugin_loaded('canada_harvest') or not plugins.plugin_loaded('harvest_portal_sync'):
+            return
+
+        registry_ini = plugins.toolkit.config.get('ckanext.canada.portal_sync.registry_ini')
+        portal_ini = plugins.toolkit.config.get('ckanext.canada.portal_sync.portal_ini')
+        if not registry_ini:
+            raise Exception('ckanext.canada.portal_sync.registry_ini not defined')
+        if not isfile(registry_ini):
+            raise Exception('Cannot find file %s, defined from ckanext.canada.portal_sync.registry_ini' % registry_ini)
+        if not portal_ini:
+            raise Exception('ckanext.canada.portal_sync.portal_ini not defined')
+        if not isfile(portal_ini):
+            raise Exception('Cannot find file %s, defined from ckanext.canada.portal_sync.portal_ini' % portal_ini)
+
+        registry_config = ConfigParser()
+        registry_config.read(registry_ini)
+        try:
+            registry_datastore_url = registry_config.get('app:main', 'ckan.datastore.write_url')
+        except NoOptionError:
+            raise Exception('Cannot find ckan.datastore.write_url in app:main section of Registry INI')
+        self.registry_datastore_url = str(_get_engine_from_url(registry_datastore_url).url)  # ensure that this is actually a DS database
+        portal_config = ConfigParser()
+        portal_config.read(portal_ini)
+        try:
+            portal_datastore_url = portal_config.get('app:main', 'ckan.datastore.write_url')
+        except NoOptionError:
+            raise Exception('Cannot find ckan.datastore.write_url in app:main section of Portal INI')
+        self.portal_datastore_url = str(_get_engine_from_url(portal_datastore_url).url)  # ensure that this is actually a DS database
+
+        #FIXME: figure out how to get the data from and to the Portal in fetch_stage and import_stage
+        self.registry = LocalCKAN()
+        self.portal = LocalCKAN()
```

Comment:
I don't think this is possible. LocalCKAN is only for the local instance of CKAN that is running, you can't have two of them with different configurations because the configuration is global.

Reply:
Yeah I am going to look into the threading library and see if that works for us at all.

Reply:
Even more chaos, I made a context manager in the class which initializes the Portal, effectively switching the entire app context and everything to the Portal. And then on exit, it will switch back to the Registry app context.

Reply:
Yikes, and that has to be really slow if it works at all.

Reply:
It does in fact work interestingly enough haha.

In the gather_stage I made sure to initiate the context manager before the looping of packages so it doesn't do it for each package. But my concern is in the import_stage as the import_stage is only passed one package at a time.

It isn't as slow as you would think, but still unsure of it. Do you have any ideas to solve this? I tried using the Thread lib but that did not really work to change to the Portal app context while still maintaining the Registry app context afterwards :/